AppCenter/AppCenter/Internals/Context/AuthToken/MSAuthTokenContext.m

```diff
@@ -42,22 +50,39 @@ + (void)resetSharedInstance {
   sharedInstance = nil;
 }
 
-- (NSString *)authToken {
+- (nullable NSString *)getAuthToken {
   @synchronized(self) {
-    return _authToken;
+    return self.authToken;
   }
 }
 
-- (void)setAuthToken:(NSString *)authToken {
+- (void)setAuthToken:(NSString *)authToken withAccountId:(NSString *)accountId {
   NSArray *synchronizedDelegates;
+  BOOL isNewUser = NO;
   @synchronized(self) {
-    _authToken = authToken;
+    self.authToken = authToken;
+    isNewUser = self.lastHomeAccountId == nil || ![self.lastHomeAccountId isEqualToString:accountId];
+    self.lastHomeAccountId = accountId;
+    
+    // Don't invoke the delegate while locking; it might be locking too and deadlock ourselves.
+    synchronizedDelegates = [self.delegates allObjects];
+  }
+  for (id<MSAuthTokenContextDelegate> delegate in synchronizedDelegates) {
+    [delegate authTokenContext:self didReceiveAuthToken:authToken forNewUser:isNewUser];
+  }
+}
 
+- (void)clearAuthToken {
+  NSArray *synchronizedDelegates;
+  @synchronized(self) {
+    self.authToken = nil;
+    self.lastHomeAccountId = nil;
+    
     // Don't invoke the delegate while locking; it might be locking too and deadlock ourselves.
     synchronizedDelegates = [self.delegates allObjects];
   }
   for (id<MSAuthTokenContextDelegate> delegate in synchronizedDelegates) {
-    [delegate authTokenContext:self didReceiveAuthToken:authToken];
+    [delegate authTokenContext:self didReceiveAuthToken:nil forNewUser:YES];
```

Comment:
Calling `YES` on `forNewUser:` delegate to trigger Push module send a new push token sounds like a confusing and misleading method name. It's very confusing without context of the problem as `forNewUser:YES` tells us it's calling delegate because of a new user, but that's not right.

Consider rename this method to make it less confusing.